### PR TITLE
Check const functions for mutable references

### LINKED ISSUE #1552: Mutable references are forbidden in const functions
I tried this code:

```rust
const fn foo(a: &mut i32) { // { dg-error "mutable references are not allowed in const functions" }
    *a = 15;
}
```

I expected to see this happen: 
```rust
error[E0658]: mutable references are not allowed in constant functions
 --> test2.rs:1:14
  |
1 | const fn foo(a: &mut i32) {
  |              ^
  |
  = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information

error: aborting due to previous error
```

Instead, this happened: compiled without issue, even when using the value.

### Meta

- What version of Rust GCC were you using: 8ff035ddc55

## Task List

- [ ] Adding two types of context in `ConstChecker`: functions/methods and const values (and const blocks later?)
- [ ] Check when a type is being visited: If we're on a mutable reference and the context is a function/method, error out

Testcase available on https://github.com/CohenArthur/gccrs/tree/forbid-mut-refs-in-const-fns

### PR DESCRIPTION
**Pull request by dme2**
Fixes #1552

### CODE REVIEW

**Comment by dkm** [comment]
Hello Dave,
Thanks for the new PR. I was checking if we had already some sign of us having a FSF copyright assignment or if you've correctly used Signed-Of to indicate you are using the DCO, but couldn't really find anything.

Can you take care of this for future contribution ? We can help if you need any help or have any question on that matter :)

**Comment by dme2** [comment]
Hi Marc, can I sign off here in the comment section? For future contributions I'll just use git to sign off.

**Comment by dkm** [comment]
You can add an explicit mention in the PR that was already merged, stating that you've read the GCC's DCO and you are using it. Then simply add the correct Sign-off-by to future commit... Should be enough!

We will need to amend your previous commit before merging it in GCC, so having an explicit ACK is probably best!

**Comment by dme2** [comment]
Just signed off on the other PR (https://github.com/Rust-GCC/gccrs/pull/1636#issue-1444430441)

**Comment by dkm** [comment]
Thanks @dme2!

File: gcc/rust/checks/errors/rust-const-checker.h [hunk 1]
```diff
@@ -71,6 +71,9 @@ class ConstChecker : public HIRFullVisitor
   void check_default_const_generics (
     std::vector<std::unique_ptr<GenericParam>> &param, ConstGenericCtx context);
 
+  /* Are we currently in a const context? */
+  bool in_const_context = false;
```
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-const-checker.h:75]: This is not necessary and also bug prone. You can read the whole discussion [here](https://github.com/Rust-GCC/gccrs/pull/1410#discussion_r930960483) which is what led us to come up with the `StackedContext` class, which you can find right under your boolean :) You can check whether or not you are in a constant context (like a constant function) by calling `const_context.is_in_context ()`, which will return a boolean.

The class is defined [here](https://github.com/Rust-GCC/gccrs/blob/master/gcc/rust/util/rust-stacked-contexts.h), in `gcc/rust/util/rust-stacked-context.h`

File: gcc/rust/checks/errors/rust-const-checker.cc [hunk 2]
```diff
@@ -898,8 +901,12 @@ ConstChecker::visit (RawPointerType &)
 {}
 
 void
-ConstChecker::visit (ReferenceType &)
-{}
+ConstChecker::visit (ReferenceType &type)
+{
+  if (this->in_const_context && type.is_mut ())
+    rust_error_at (type.get_locus (),
+		   "Mutable references are not allowed in constant functions");
```
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-const-checker.cc:908]:
```suggestion
		   "mutable references are not allowed in constant functions");
```

We usually go for lowercase error strings as it looks better when formatted by GCC

```shell
> gccrs
gccrs: fatal error: no input files
compilation terminated.
```

File: gcc/rust/checks/errors/rust-const-checker.cc [hunk 3]
```diff
@@ -898,8 +901,12 @@ ConstChecker::visit (RawPointerType &)
 {}
 
 void
-ConstChecker::visit (ReferenceType &)
-{}
+ConstChecker::visit (ReferenceType &type)
+{
+  if (this->in_const_context && type.is_mut ())
```
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-const-checker.cc:906]:
```suggestion
  if (const_context.is_in_context () && type.is_mut ())
```

File: gcc/rust/checks/errors/rust-const-checker.cc [hunk 4]
```diff
@@ -617,7 +617,10 @@ ConstChecker::visit (Function &function)
 {
   auto const_fn = function.get_qualifiers ().is_const ();
   if (const_fn)
-    const_context.enter (function.get_mappings ().get_hirid ());
+    {
+      const_context.enter (function.get_mappings ().get_hirid ());
+      this->in_const_context = true;
```
Comment (CohenArthur) [inline gcc/rust/checks/errors/rust-const-checker.cc:622]: So yeah, no need for that line :) also, you were never setting it back to false, so anytime we'd see one constant function we'd be in a constant context for the rest of the crate we're checking

**Review by CohenArthur — COMMENTED** [review]
Thank you for contributing and taking the time to sign off @dme2! I'm sorry to be so annoying on which API to use :) let me know if you think the `StackedContext` API deservers improvements or more documentation overall, as I understand it might be quite unclear at first. Thanks again!

**Comment by dme2** [comment]
Thank you for reviewing that, @CohenArthur .The StackedContext API is pretty much exactly what I was looking for! I went ahead and made those recommended changes.

**Review by CohenArthur — APPROVED** [review]
Thank you :) Looks great!

bors r+

**Comment by CohenArthur** [comment]
bors cancel

Oops one more thing. @dme2 could you rebase all those commits in a single commit?

**Comment by dme2** [comment]
I think I did that rebase correctly. If more changes are necessary just let me know

**Comment by CohenArthur** [comment]
Thanks @dme2 that's great :)

bors r+

**Pull request merged**